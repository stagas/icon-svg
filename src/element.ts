import { getIconUrl, fetchIconSvg, propsSatisfyTemplate } from '.'

export type HTMLIconSvgElement = Parameters<typeof getIconUrl>[0] & {
  raw?: boolean
  'stroke-width'?: string | number | boolean
}

declare global {
  interface CSSStyleSheet {
    replace(css: string): Promise<void>
    replaceSync(css: string): void
  }
  interface ShadowRoot {
    adoptedStyleSheets: readonly CSSStyleSheet[]
  }
}

const style = new CSSStyleSheet()

const replaceStyle = ({ strokeWidth }: { strokeWidth: string | number | null }) => {
  style.replaceSync(/*css*/ `
    :host {
      ${strokeWidth ? `stroke-width: ${strokeWidth};` : ''}
      --stroke: currentColor;
      --fill: #9994;
      --fill-secondary: #9996;
      display: inline-flex;
      vertical-align: middle;
    }

    [part=svg] {
      width: 100%;
      height: 100%;
    }
  `)
}

/**
 * The `IconSvgElement` custom element.
 *
 * For a more complete example with typings check `playground/app.tsx`
 *
 * Browser needs to support `Constructible StyleSheets` ([polyfill](https://github.com/calebdwilliams/construct-style-sheets)).
 *
 * _JS / TS:_
 * ```ts
 * customElements.define('icon-svg', IconSvgElement)
 * ```
 *
 * _CSS:_
 * ```css
 * icon-svg {
 *   width: 32px;
 *   height: 32px;
 *   stroke-width: 2.5px;
 *   --stroke: #fff;
 *   --fill: #028;
 *   --fill-secondary: #468;
 * }
 * ```
 *
 * _HTML:_
 * ```html
 * <icon-svg set="boxicons" type="logos" icon="javascript"></icon-svg>
 * ```
 */
export class IconSvgElement extends HTMLElement {
  static get observedAttributes() {
    return ['icon', 'set', 'type', 'kind', 'raw', 'stroke-width']
  }

  constructor() {
    super()
    this.attachShadow({ mode: 'open' }).adoptedStyleSheets = [style]
    this.updateSvg()
  }

  async updateSvg() {
    const props = {
      set: this.getAttribute('set'),
      icon: this.getAttribute('icon'),
      type: this.getAttribute('type'),
      kind: this.getAttribute('kind'),
      raw: this.hasAttribute('raw'),
      strokeWidth: this.getAttribute('stroke-width'),
    }
    if (!propsSatisfyTemplate(props)) return

    replaceStyle(props)
    this.shadowRoot!.innerHTML = (await fetchIconSvg(props as never)).replace(/(<svg.*?)>/, '$1 part="svg">')
  }

  attributeChangedCallback() {
    this.updateSvg()
  }
}
