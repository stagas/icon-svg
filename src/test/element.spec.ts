import 'construct-style-sheets-polyfill'
import { IconSvgElement } from '../element'

describe('IconSvgElement', () => {
  it('registers', () => {
    customElements.define('icon-svg', IconSvgElement)
    expect(customElements.get('icon-svg')).toBe(IconSvgElement)
  })

  it('can be created', () => {
    const iconSvg = document.createElement('icon-svg') as IconSvgElement
    expect(iconSvg).toBeInstanceOf(IconSvgElement)
  })

  it('creates icon svg when it has the necessary attributes', done => {
    let fetchUrl: string
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ;(window as any).fetch = async (url: string) => {
      fetchUrl = url
      return {
        ok: true,
        text: async () => `<svg><path stroke-width="1.33" /></svg>`,
      }
    }
    const iconSvg = document.createElement('icon-svg') as IconSvgElement
    iconSvg.setAttribute('set', 'feather')
    iconSvg.setAttribute('icon', 'award')
    iconSvg.setAttribute('stroke-width', '2px')
    queueMicrotask(() => {
      expect(fetchUrl).toMatchSnapshot()
      done()
    })
  })
})
