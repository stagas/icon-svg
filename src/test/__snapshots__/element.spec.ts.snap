// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSvgElement creates icon svg when it has the necessary attributes 1`] = `"https://cdn.jsdelivr.net/npm/feather-icons@4/dist/icons/award.svg"`;
